File: Laborator/Java/AgentieTurism/src/main/java/ro/mpp2024/repository/EmployeeDBRepository.java
package ro.mpp2024.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.domain.Employee;
import ro.mpp2024.utils.JdbcUtils;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.sql.*;


public class EmployeeDBRepository implements EmployeeRepository {
    private JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public EmployeeDBRepository(Properties props) {
        logger.info("Initializing EmployeeDBRepository with properties: {}", props);
        dbUtils = new JdbcUtils(props);
    }

    public static String hashPassword(String password) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] encodedHash = digest.digest(password.getBytes(StandardCharsets.UTF_8));

            StringBuilder hexString = new StringBuilder();
            for (byte b : encodedHash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }

            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Error hashing password", e);
        }
    }

    public void add(Employee employee) {
        logger.traceEntry("Saving employee {}", employee);
        String sql = "INSERT INTO Employees (name, email, password) VALUES (?, ?, ?)";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            stmt.setString(1, employee.getName());
            stmt.setString(2, employee.getEmail());
            stmt.setString(3, hashPassword(employee.getPassword()));

            int affectedRows = stmt.executeUpdate();
            if (affectedRows == 0) {
                throw new SQLException("Creating employee failed, no rows affected.");
            }

            try (ResultSet generatedKeys = stmt.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    int generatedId = generatedKeys.getInt(1);
                    employee.setId(generatedId);
                    logger.trace("Employee saved successfully with ID: {}", generatedId);
                } else {
                    throw new SQLException("Creating employee failed, no ID obtained.");
                }
            }
        } catch (SQLException e) {
            logger.error("Error inserting employee", e);
        }
        logger.traceExit();
    }


    @Override
    public void update(Integer id, Employee employee) {
        logger.traceEntry("Updating employee with ID: {} to {}", id, employee);
        String sql = "UPDATE Employees SET name = ?, email = ?, password = ? WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setString(1, employee.getName());
            stmt.setString(2, employee.getEmail());
            stmt.setString(3, hashPassword(employee.getPassword()));
            stmt.setInt(4, id);
            int result = stmt.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer id) {
        logger.traceEntry("Deleting employee with ID: {}", id);
        String sql = "DELETE FROM Employees WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setInt(1, id);
            int result = stmt.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public Employee findOne(Integer id) {
        logger.traceEntry("Finding employee by ID: {}", id);
        String sql = "SELECT * FROM Employees WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setInt(1, id);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Employee employee = new Employee(rs.getString("name"), rs.getString("email"), rs.getString("password"));
                employee.setId(rs.getInt("id"));
                logger.traceExit(employee);
                return employee;
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        return null;
    }


    @Override
    public Employee logIn(String email, String password) {
        logger.traceEntry("Finding employee by email: {}", email);
        String sql = "SELECT * FROM Employees WHERE email = ? AND password = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setString(1, email);
            stmt.setString(2, hashPassword(password));
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Employee employee = new Employee(rs.getString("name"), rs.getString("email"), rs.getString("password"));
                employee.setId(rs.getInt("id"));
                logger.traceExit(employee);
                return employee;
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        return null;
    }

    @Override
    public Iterable<Employee> findAll() {
        logger.traceEntry();
        List<Employee> employees = new ArrayList<>();
        String sql = "SELECT * FROM Employees";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql);
             ResultSet rs = stmt.executeQuery()) {
            while (rs.next()) {
                Employee employee = new Employee(rs.getString("name"), rs.getString("email"), rs.getString("password"));
                employee.setId(rs.getInt("id"));
                employees.add(employee);
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit(employees);
        return employees;
    }
}

File: Laborator/CSharp/AgentieTurism/Repository/EmployeeDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using AgentieTurism.Models;
using AgentieTurism.Utils;
using log4net;

namespace AgentieTurism.Repository
{
    public class EmployeeDbRepository : IEmployeeRepository
    {
        private static readonly ILog log = LogManager.GetLogger("EmployeeDbRepo");
        //private readonly IDictionary<string, string> _properties;

        //public EmployeeDbRepository(IDictionary<string, string> props)
        public EmployeeDbRepository()
        {
            log.Info("Initializing EmployeeDbRepository");
            //_properties = props;
        }

        public void Add(Employee employee)
        {
            log.Info($"Adding employee: {employee}");
            string hashedPassword = HashPassword(employee.Password);

            var sql = "INSERT INTO Employees (name, email, password) VALUES (@Name, @Email, @Password)";
            var parameters = new Dictionary<string, object>
            {
                { "@Name", employee.Name },
                { "@Email", employee.Email },
                { "@Password", hashedPassword }
            };

            try
            {
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    foreach (var param in parameters)
                    {
                        var dbParam = cmd.CreateParameter();
                        dbParam.ParameterName = param.Key;
                        dbParam.Value = param.Value;
                        cmd.Parameters.Add(dbParam);
                    }
                    cmd.ExecuteNonQuery();

                    cmd.CommandText = "SELECT last_insert_rowid()";
                    employee.Id = Convert.ToInt32(cmd.ExecuteScalar());
                }

                log.Info($"Employee {employee.Name} added successfully");
            }
            catch (Exception ex)
            {
                log.Error("Failed to add employee", ex);
            }
        }

        public void Update(int id, Employee employee)
        {
            log.Info($"Updating employee with ID: {id}");
            string hashedPassword = HashPassword(employee.Password);

            var sql = "UPDATE Employees SET name = @Name, email = @Email, password = @Password WHERE id = @Id";
            var parameters = new Dictionary<string, object>
            {
                { "@Id", id },
                { "@Name", employee.Name },
                { "@Email", employee.Email },
                { "@Password", hashedPassword }
            };
            ExecuteNonQuery(sql, parameters);
            log.Info($"Employee with ID: {id} updated successfully");
        }

        public void Delete(int id)
        {
            log.Info($"Deleting employee with ID: {id}");
            var sql = "DELETE FROM Employees WHERE id = @Id";
            var parameters = new Dictionary<string, object> { { "@Id", id } };
            ExecuteNonQuery(sql, parameters);
            log.Info($"Employee with ID: {id} deleted successfully");
        }

        public Employee FindOne(int id)
        {
            log.Info($"Finding employee by ID: {id}");
            var sql = "SELECT * FROM Employees WHERE id = @Id";
            var parameters = new Dictionary<string, object> { { "@Id", id } };
            var employee = ExecuteQuerySingle(sql, parameters, DecodeEmployee);
            log.Info(employee != null ? $"Employee found: {employee}" : "Employee not found");
            return employee;
        }

        public List<Employee> FindAll()
        {
            log.Info("Retrieving all employees");
            var sql = "SELECT * FROM Employees";
            var employees = ExecuteQueryList(sql, null, DecodeEmployee);
            log.Info($"Found {employees.Count} employees");
            return employees;
        }

        public Employee LogIn(string email, string password)
        {
            log.Info($"Attempting login for email: {email}");
            string hashedPassword = HashPassword(password);

            var sql = "SELECT * FROM Employees WHERE email = @Email AND password = @Password";
            var parameters = new Dictionary<string, object>
            {
                { "@Email", email },
                { "@Password", hashedPassword }
            };

            var employee = ExecuteQuerySingle(sql, parameters, DecodeEmployee);
            log.Info(employee != null ? "Login successful" : "Login failed");
            return employee;
        }

        private string HashPassword(string password)
        {
            using (SHA256 sha256 = SHA256.Create())
            {
                byte[] bytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(password));
                StringBuilder builder = new StringBuilder();
                foreach (byte b in bytes)
                {
                    builder.Append(b.ToString("x2"));
                }
                return builder.ToString();
            }
        }

        private Employee DecodeEmployee(IDataReader reader)
        {
            Employee employee = new Employee(
                reader.GetString(1),
                reader.GetString(2),
                reader.GetString(3)
            );
            employee.Id = reader.GetInt32(0);

            return employee;
        }

        private void ExecuteNonQuery(string sql, Dictionary<string, object> parameters)
        {
            try
            {
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    foreach (var param in parameters)
                    {
                        var dbParam = cmd.CreateParameter();
                        dbParam.ParameterName = param.Key;
                        dbParam.Value = param.Value;
                        cmd.Parameters.Add(dbParam);
                    }
                    cmd.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                log.Error("Database operation failed", ex);
            }
        }

        private Employee ExecuteQuerySingle(string sql, Dictionary<string, object> parameters, Func<IDataReader, Employee> decoder)
        {
            try
            {
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    if (parameters != null)
                    {
                        foreach (var param in parameters)
                        {
                            var dbParam = cmd.CreateParameter();
                            dbParam.ParameterName = param.Key;
                            dbParam.Value = param.Value;
                            cmd.Parameters.Add(dbParam);
                        }
                    }

                    using (var reader = cmd.ExecuteReader())
                    {
                        return reader.Read() ? decoder(reader) : null;
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Database query failed", ex);
                return null;
            }
        }

        private List<Employee> ExecuteQueryList(string sql, Dictionary<string, object> parameters, Func<IDataReader, Employee> decoder)
        {
            try
            {
                var employees = new List<Employee>();
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    if (parameters != null)
                    {
                        foreach (var param in parameters)
                        {
                            var dbParam = cmd.CreateParameter();
                            dbParam.ParameterName = param.Key;
                            dbParam.Value = param.Value;
                            cmd.Parameters.Add(dbParam);
                        }
                    }

                    using (var reader = cmd.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            employees.Add(decoder(reader));
                        }
                    }
                }
                return employees;
            }
            catch (Exception ex)
            {
                log.Error("Database query failed", ex);
                return new List<Employee>();
            }
        }
    }
}
